File: org.kevoree.microsandbox.watchdog/scratch/InstrTreeWalker.cpp
/* $ANTLR 2.7.2: "instr.g" -> "InstrTreeWalker.cpp"$ */
#include "InstrTreeWalker.hpp"
#include <antlr/Token.hpp>
#include <antlr/AST.hpp>
#include <antlr/NoViableAltException.hpp>
#include <antlr/MismatchedTokenException.hpp>
#include <antlr/SemanticException.hpp>
#include <antlr/BitSet.hpp>
#line 1 "instr.g"
#line 11 "InstrTreeWalker.cpp"
InstrTreeWalker::InstrTreeWalker()
	: ANTLR_USE_NAMESPACE(antlr)TreeParser() {
}

void InstrTreeWalker::slist(ANTLR_USE_NAMESPACE(antlr)RefAST _t) {
	ANTLR_USE_NAMESPACE(antlr)RefAST slist_AST_in = _t;
	
	try {      // for error handling
#line 117 "instr.g"
		dumpHidden(filter->getInitialHiddenToken());
#line 22 "InstrTreeWalker.cpp"
		{ // ( ... )+
		int _cnt21081=0;
		for (;;) {
			if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
				_t = ASTNULL;
			if ((_tokenSet_0.member(_t->getType()))) {
				stat(_t);
				_t = _retTree;
			}
			else {
				if ( _cnt21081>=1 ) { goto _loop21081; } else {throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);}
			}
			
			_cnt21081++;
		}
		_loop21081:;
		}  // ( ... )+
	}
	catch (ANTLR_USE_NAMESPACE(antlr)RecognitionException& ex) {
		reportError(ex);
		if ( _t != ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = _t->getNextSibling();
	}
	_retTree = _t;
}

void InstrTreeWalker::stat(ANTLR_USE_NAMESPACE(antlr)RefAST _t) {
	ANTLR_USE_NAMESPACE(antlr)RefAST stat_AST_in = _t;
	ANTLR_USE_NAMESPACE(antlr)RefAST i = ANTLR_USE_NAMESPACE(antlr)nullAST;
	ANTLR_USE_NAMESPACE(antlr)RefAST t = ANTLR_USE_NAMESPACE(antlr)nullAST;
	ANTLR_USE_NAMESPACE(antlr)RefAST e = ANTLR_USE_NAMESPACE(antlr)nullAST;
	
	try {      // for error handling
		if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = ASTNULL;
		switch ( _t->getType()) {
		case LBRACE:
		{
			ANTLR_USE_NAMESPACE(antlr)RefAST __t21083 = _t;
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp17_AST_in = _t;
			match(_t,LBRACE);
			_t = _t->getFirstChild();
#line 121 "instr.g"
			pr(tmp17_AST_in);
#line 67 "InstrTreeWalker.cpp"
			{ // ( ... )+
			int _cnt21085=0;
			for (;;) {
				if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
					_t = ASTNULL;
				if ((_tokenSet_0.member(_t->getType()))) {
					stat(_t);
					_t = _retTree;
				}
				else {
					if ( _cnt21085>=1 ) { goto _loop21085; } else {throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);}
				}
				
				_cnt21085++;
			}
			_loop21085:;
			}  // ( ... )+
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp18_AST_in = _t;
			match(_t,RBRACE);
			_t = _t->getNextSibling();
#line 121 "instr.g"
			pr(tmp18_AST_in);
#line 90 "InstrTreeWalker.cpp"
			_t = __t21083;
			_t = _t->getNextSibling();
			break;
		}
		case TK_if:
		{
			ANTLR_USE_NAMESPACE(antlr)RefAST __t21086 = _t;
			i = (_t == ASTNULL) ? ANTLR_USE_NAMESPACE(antlr)nullAST : _t;
			match(_t,TK_if);
			_t = _t->getFirstChild();
#line 122 "instr.g"
			pr(i);
#line 103 "InstrTreeWalker.cpp"
			expr(_t);
			_t = _retTree;
			t = _t;
			match(_t,TK_then);
			_t = _t->getNextSibling();
#line 122 "instr.g"
			pr(t);
#line 111 "InstrTreeWalker.cpp"
			stat(_t);
			_t = _retTree;
			{
			if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
				_t = ASTNULL;
			switch ( _t->getType()) {
			case TK_else:
			{
				e = _t;
				match(_t,TK_else);
				_t = _t->getNextSibling();
#line 122 "instr.g"
				pr(e);
#line 125 "InstrTreeWalker.cpp"
				stat(_t);
				_t = _retTree;
				break;
			}
			case 3:
			{
				break;
			}
			default:
			{
				throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);
			}
			}
			}
			_t = __t21086;
			_t = _t->getNextSibling();
			break;
		}
		case ASSIGN:
		{
			ANTLR_USE_NAMESPACE(antlr)RefAST __t21088 = _t;
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp19_AST_in = _t;
			match(_t,ASSIGN);
			_t = _t->getFirstChild();
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp20_AST_in = _t;
			match(_t,ID);
			_t = _t->getNextSibling();
#line 123 "instr.g"
			pr(tmp20_AST_in); pr(tmp19_AST_in);
#line 155 "InstrTreeWalker.cpp"
			expr(_t);
			_t = _retTree;
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp21_AST_in = _t;
			match(_t,SEMI);
			_t = _t->getNextSibling();
#line 123 "instr.g"
			pr(tmp21_AST_in);
#line 163 "InstrTreeWalker.cpp"
			_t = __t21088;
			_t = _t->getNextSibling();
			break;
		}
		case CALL:
		{
			call(_t);
			_t = _retTree;
			break;
		}
		default:
		{
			throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);
		}
		}
	}
	catch (ANTLR_USE_NAMESPACE(antlr)RecognitionException& ex) {
		reportError(ex);
		if ( _t != ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = _t->getNextSibling();
	}
	_retTree = _t;
}

void InstrTreeWalker::expr(ANTLR_USE_NAMESPACE(antlr)RefAST _t) {
	ANTLR_USE_NAMESPACE(antlr)RefAST expr_AST_in = _t;
	
	try {      // for error handling
		if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = ASTNULL;
		switch ( _t->getType()) {
		case PLUS:
		{
			ANTLR_USE_NAMESPACE(antlr)RefAST __t21090 = _t;
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp22_AST_in = _t;
			match(_t,PLUS);
			_t = _t->getFirstChild();
			expr(_t);
			_t = _retTree;
#line 128 "instr.g"
			pr(tmp22_AST_in);
#line 205 "InstrTreeWalker.cpp"
			expr(_t);
			_t = _retTree;
			_t = __t21090;
			_t = _t->getNextSibling();
			break;
		}
		case STAR:
		{
			ANTLR_USE_NAMESPACE(antlr)RefAST __t21091 = _t;
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp23_AST_in = _t;
			match(_t,STAR);
			_t = _t->getFirstChild();
			expr(_t);
			_t = _retTree;
#line 129 "instr.g"
			pr(tmp23_AST_in);
#line 222 "InstrTreeWalker.cpp"
			expr(_t);
			_t = _retTree;
			_t = __t21091;
			_t = _t->getNextSibling();
			break;
		}
		case INT:
		{
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp24_AST_in = _t;
			match(_t,INT);
			_t = _t->getNextSibling();
#line 130 "instr.g"
			pr(tmp24_AST_in);
#line 236 "InstrTreeWalker.cpp"
			break;
		}
		case ID:
		{
			ANTLR_USE_NAMESPACE(antlr)RefAST tmp25_AST_in = _t;
			match(_t,ID);
			_t = _t->getNextSibling();
#line 131 "instr.g"
			pr(tmp25_AST_in);
#line 246 "InstrTreeWalker.cpp"
			break;
		}
		default:
		{
			throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);
		}
		}
	}
	catch (ANTLR_USE_NAMESPACE(antlr)RecognitionException& ex) {
		reportError(ex);
		if ( _t != ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = _t->getNextSibling();
	}
	_retTree = _t;
}

void InstrTreeWalker::call(ANTLR_USE_NAMESPACE(antlr)RefAST _t) {
	ANTLR_USE_NAMESPACE(antlr)RefAST call_AST_in = _t;
	
	try {      // for error handling
#line 134 "instr.g"
		
				// add instrumentation about call; manually call rule
				callDumpInstrumentation(call_AST_in);
				
#line 272 "InstrTreeWalker.cpp"
		ANTLR_USE_NAMESPACE(antlr)RefAST __t21093 = _t;
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp26_AST_in = _t;
		match(_t,CALL);
		_t = _t->getFirstChild();
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp27_AST_in = _t;
		match(_t,ID);
		_t = _t->getNextSibling();
#line 138 "instr.g"
		pr(tmp27_AST_in);
#line 282 "InstrTreeWalker.cpp"
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp28_AST_in = _t;
		match(_t,LPAREN);
		_t = _t->getNextSibling();
#line 139 "instr.g"
		pr(tmp28_AST_in);
#line 288 "InstrTreeWalker.cpp"
		{
		if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = ASTNULL;
		switch ( _t->getType()) {
		case ID:
		case PLUS:
		case STAR:
		case INT:
		{
			expr(_t);
			_t = _retTree;
			break;
		}
		case RPAREN:
		{
			break;
		}
		default:
		{
			throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);
		}
		}
		}
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp29_AST_in = _t;
		match(_t,RPAREN);
		_t = _t->getNextSibling();
#line 139 "instr.g"
		pr(tmp29_AST_in);
#line 317 "InstrTreeWalker.cpp"
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp30_AST_in = _t;
		match(_t,SEMI);
		_t = _t->getNextSibling();
#line 141 "instr.g"
		
				  // print SEMI manually; need '}' between it and whitespace
				  std::cout << tmp30_AST_in->getText();
				  std::cout << "}"; // close {...} of instrumentation
				  dumpHidden(
					(ANTLR_USE_NAMESPACE(antlr)RefCommonASTWithHiddenTokens(tmp30_AST_in))->getHiddenAfter()
				  );
				
#line 330 "InstrTreeWalker.cpp"
		_t = __t21093;
		_t = _t->getNextSibling();
	}
	catch (ANTLR_USE_NAMESPACE(antlr)RecognitionException& ex) {
		reportError(ex);
		if ( _t != ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = _t->getNextSibling();
	}
	_retTree = _t;
}

/** Dump instrumentation for a call statement.
 *  The reference to rule expr prints out the arg
 *  and then at the end of this rule, we close the
 *  generated called to dbg.invoke().
 */
void InstrTreeWalker::callDumpInstrumentation(ANTLR_USE_NAMESPACE(antlr)RefAST _t) {
	ANTLR_USE_NAMESPACE(antlr)RefAST callDumpInstrumentation_AST_in = _t;
	ANTLR_USE_NAMESPACE(antlr)RefAST id = ANTLR_USE_NAMESPACE(antlr)nullAST;
	ANTLR_USE_NAMESPACE(antlr)RefAST e = ANTLR_USE_NAMESPACE(antlr)nullAST;
	
	try {      // for error handling
		ANTLR_USE_NAMESPACE(antlr)RefAST __t21096 = _t;
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp31_AST_in = _t;
		match(_t,CALL);
		_t = _t->getFirstChild();
		id = _t;
		match(_t,ID);
		_t = _t->getNextSibling();
#line 159 "instr.g"
		std::cout << "{dbg.invoke(\"" << id->getText() << "\", \"";
#line 362 "InstrTreeWalker.cpp"
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp32_AST_in = _t;
		match(_t,LPAREN);
		_t = _t->getNextSibling();
		{
		if (_t == ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = ASTNULL;
		switch ( _t->getType()) {
		case ID:
		case PLUS:
		case STAR:
		case INT:
		{
			e = (_t == ASTNULL) ? ANTLR_USE_NAMESPACE(antlr)nullAST : _t;
			expr(_t);
			_t = _retTree;
			break;
		}
		case RPAREN:
		{
			break;
		}
		default:
		{
			throw ANTLR_USE_NAMESPACE(antlr)NoViableAltException(_t);
		}
		}
		}
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp33_AST_in = _t;
		match(_t,RPAREN);
		_t = _t->getNextSibling();
		ANTLR_USE_NAMESPACE(antlr)RefAST tmp34_AST_in = _t;
		match(_t,SEMI);
		_t = _t->getNextSibling();
#line 161 "instr.g"
		std::cout << "\"); ";
#line 398 "InstrTreeWalker.cpp"
		_t = __t21096;
		_t = _t->getNextSibling();
	}
	catch (ANTLR_USE_NAMESPACE(antlr)RecognitionException& ex) {
		reportError(ex);
		if ( _t != ANTLR_USE_NAMESPACE(antlr)nullAST )
			_t = _t->getNextSibling();
	}
	_retTree = _t;
}

void InstrTreeWalker::initializeASTFactory( ANTLR_USE_NAMESPACE(antlr)ASTFactory& factory )
{
}
const char* InstrTreeWalker::tokenNames[] = {
	"<0>",
	"EOF",
	"<2>",
	"NULL_TREE_LOOKAHEAD",
	"CALL",
	"LBRACE",
	"RBRACE",
	"\"if\"",
	"\"then\"",
	"\"else\"",
	"ID",
	"ASSIGN",
	"SEMI",
	"PLUS",
	"STAR",
	"INT",
	"LPAREN",
	"RPAREN",
	"WS_",
	"SL_COMMENT",
	"DIGIT",
	0
};

const unsigned long InstrTreeWalker::_tokenSet_0_data_[] = { 2224UL, 0UL, 0UL, 0UL };
// CALL LBRACE "if" ASSIGN 
const ANTLR_USE_NAMESPACE(antlr)BitSet InstrTreeWalker::_tokenSet_0(_tokenSet_0_data_,4);

File: org.kevoree.microsandbox.watchdog/scratch/InstrTreeWalker.cs
// $ANTLR 2.7.2: "instr.g" -> "InstrTreeWalker.cs"$

	// Generate header specific to the tree-parser CSharp file
	using System;
	
	using TreeParser = antlr.TreeParser;
	using Token                    = antlr.Token;
	using AST                      = antlr.collections.AST;
	using RecognitionException     = antlr.RecognitionException;
	using ANTLRException           = antlr.ANTLRException;
	using NoViableAltException     = antlr.NoViableAltException;
	using MismatchedTokenException = antlr.MismatchedTokenException;
	using SemanticException        = antlr.SemanticException;
	using BitSet                   = antlr.collections.impl.BitSet;
	using ASTPair                  = antlr.ASTPair;
	using ASTFactory               = antlr.ASTFactory;
	using ASTArray                 = antlr.collections.impl.ASTArray;
	
	
	public 	class InstrTreeWalker : antlr.TreeParser
	{
		public const int EOF = 1;
		public const int NULL_TREE_LOOKAHEAD = 3;
		public const int CALL = 4;
		public const int LBRACE = 5;
		public const int RBRACE = 6;
		public const int TK_if = 7;
		public const int TK_then = 8;
		public const int TK_else = 9;
		public const int ID = 10;
		public const int ASSIGN = 11;
		public const int SEMI = 12;
		public const int PLUS = 13;
		public const int STAR = 14;
		public const int INT = 15;
		public const int LPAREN = 16;
		public const int RPAREN = 17;
		public const int WS = 18;
		public const int SL_COMMENT = 19;
		public const int DIGIT = 20;
		
		
	/** walk list of hidden tokens in order, printing them out */
	public static void dumpHidden(antlr.CommonHiddenStreamToken t) {
	  for ( ; t!=null ; t=InstrMain.filter.getHiddenAfter(t) ) {
	    Console.Error.Write(t.getText());
	  }
	}

	private void pr(AST p) {
		Console.Out.Write(p.getText());
		dumpHidden(
			((antlr.CommonASTWithHiddenTokens)p).getHiddenAfter()
		);
	}
		public InstrTreeWalker()
		{
			tokenNames = tokenNames_;
		}
		
	public void slist(AST _t) //throws RecognitionException
{
		
		AST slist_AST_in = (AST)_t;
		
		try {      // for error handling
			dumpHidden(InstrMain.filter.getInitialHiddenToken());
			{ // ( ... )+
			int _cnt22400=0;
			for (;;)
			{
				if (_t == null)
					_t = ASTNULL;
				if ((tokenSet_0_.member(_t.Type)))
				{
					stat(_t);
					_t = retTree_;
				}
				else
				{
					if (_cnt22400 >= 1) { goto _loop22400_breakloop; } else { throw new NoViableAltException(_t);; }
				}
				
				_cnt22400++;
			}
_loop22400_breakloop:			;
			}    // ( ... )+
		}
		catch (RecognitionException ex)
		{
			reportError(ex);
			if (null != _t)
			{
				_t = _t.getNextSibling();
			}
		}
		retTree_ = _t;
	}
	
	public void stat(AST _t) //throws RecognitionException
{
		
		AST stat_AST_in = (AST)_t;
		AST i = null;
		AST t = null;
		AST e = null;
		
		try {      // for error handling
			if (null == _t)
				_t = ASTNULL;
			switch ( _t.Type )
			{
			case LBRACE:
			{
				AST __t22402 = _t;
				AST tmp17_AST_in = (AST)_t;
				match(_t,LBRACE);
				_t = _t.getFirstChild();
				pr(tmp17_AST_in);
				{ // ( ... )+
				int _cnt22404=0;
				for (;;)
				{
					if (_t == null)
						_t = ASTNULL;
					if ((tokenSet_0_.member(_t.Type)))
					{
						stat(_t);
						_t = retTree_;
					}
					else
					{
						if (_cnt22404 >= 1) { goto _loop22404_breakloop; } else { throw new NoViableAltException(_t);; }
					}
					
					_cnt22404++;
				}
_loop22404_breakloop:				;
				}    // ( ... )+
				AST tmp18_AST_in = (AST)_t;
				match(_t,RBRACE);
				_t = _t.getNextSibling();
				pr(tmp18_AST_in);
				_t = __t22402;
				_t = _t.getNextSibling();
				break;
			}
			case TK_if:
			{
				AST __t22405 = _t;
				i = (ASTNULL == _t) ? null : (AST)_t;
				match(_t,TK_if);
				_t = _t.getFirstChild();
				pr(i);
				expr(_t);
				_t = retTree_;
				t = (AST)_t;
				match(_t,TK_then);
				_t = _t.getNextSibling();
				pr(t);
				stat(_t);
				_t = retTree_;
				{
					if (null == _t)
						_t = ASTNULL;
					switch ( _t.Type )
					{
					case TK_else:
					{
						e = (AST)_t;
						match(_t,TK_else);
						_t = _t.getNextSibling();
						pr(e);
						stat(_t);
						_t = retTree_;
						break;
					}
					case 3:
					{
						break;
					}
					default:
					{
						throw new NoViableAltException(_t);
					}
					 }
				}
				_t = __t22405;
				_t = _t.getNextSibling();
				break;
			}
			case ASSIGN:
			{
				AST __t22407 = _t;
				AST tmp19_AST_in = (AST)_t;
				match(_t,ASSIGN);
				_t = _t.getFirstChild();
				AST tmp20_AST_in = (AST)_t;
				match(_t,ID);
				_t = _t.getNextSibling();
				pr(tmp20_AST_in); pr(tmp19_AST_in);
				expr(_t);
				_t = retTree_;
				AST tmp21_AST_in = (AST)_t;
				match(_t,SEMI);
				_t = _t.getNextSibling();
				pr(tmp21_AST_in);
				_t = __t22407;
				_t = _t.getNextSibling();
				break;
			}
			case CALL:
			{
				call(_t);
				_t = retTree_;
				break;
			}
			default:
			{
				throw new NoViableAltException(_t);
			}
			 }
		}
		catch (RecognitionException ex)
		{
			reportError(ex);
			if (null != _t)
			{
				_t = _t.getNextSibling();
			}
		}
		retTree_ = _t;
	}
	
	public void expr(AST _t) //throws RecognitionException
{
		
		AST expr_AST_in = (AST)_t;
		
		try {      // for error handling
			if (null == _t)
				_t = ASTNULL;
			switch ( _t.Type )
			{
			case PLUS:
			{
				AST __t22409 = _t;
				AST tmp22_AST_in = (AST)_t;
				match(_t,PLUS);
				_t = _t.getFirstChild();
				expr(_t);
				_t = retTree_;
				pr(tmp22_AST_in);
				expr(_t);
				_t = retTree_;
				_t = __t22409;
				_t = _t.getNextSibling();
				break;
			}
			case STAR:
			{
				AST __t22410 = _t;
				AST tmp23_AST_in = (AST)_t;
				match(_t,STAR);
				_t = _t.getFirstChild();
				expr(_t);
				_t = retTree_;
				pr(tmp23_AST_in);
				expr(_t);
				_t = retTree_;
				_t = __t22410;
				_t = _t.getNextSibling();
				break;
			}
			case INT:
			{
				AST tmp24_AST_in = (AST)_t;
				match(_t,INT);
				_t = _t.getNextSibling();
				pr(tmp24_AST_in);
				break;
			}
			case ID:
			{
				AST tmp25_AST_in = (AST)_t;
				match(_t,ID);
				_t = _t.getNextSibling();
				pr(tmp25_AST_in);
				break;
			}
			default:
			{
				throw new NoViableAltException(_t);
			}
			 }
		}
		catch (RecognitionException ex)
		{
			reportError(ex);
			if (null != _t)
			{
				_t = _t.getNextSibling();
			}
		}
		retTree_ = _t;
	}
	
	public void call(AST _t) //throws RecognitionException
{
		
		AST call_AST_in = (AST)_t;
		
		try {      // for error handling
			
					// add instrumentation about call; manually call rule
					callDumpInstrumentation(call_AST_in);
					
			AST __t22412 = _t;
			AST tmp26_AST_in = (AST)_t;
			match(_t,CALL);
			_t = _t.getFirstChild();
			AST tmp27_AST_in = (AST)_t;
			match(_t,ID);
			_t = _t.getNextSibling();
			pr(tmp27_AST_in);
			AST tmp28_AST_in = (AST)_t;
			match(_t,LPAREN);
			_t = _t.getNextSibling();
			pr(tmp28_AST_in);
			{
				if (null == _t)
					_t = ASTNULL;
				switch ( _t.Type )
				{
				case ID:
				case PLUS:
				case STAR:
				case INT:
				{
					expr(_t);
					_t = retTree_;
					break;
				}
				case RPAREN:
				{
					break;
				}
				default:
				{
					throw new NoViableAltException(_t);
				}
				 }
			}
			AST tmp29_AST_in = (AST)_t;
			match(_t,RPAREN);
			_t = _t.getNextSibling();
			pr(tmp29_AST_in);
			AST tmp30_AST_in = (AST)_t;
			match(_t,SEMI);
			_t = _t.getNextSibling();
			
					  // print SEMI manually; need '}' between it and whitespace
					  Console.Error.Write(tmp30_AST_in.getText());
					  Console.Error.Write("}"); // close {...} of instrumentation
					  dumpHidden(
						((antlr.CommonASTWithHiddenTokens)tmp30_AST_in).getHiddenAfter()
					  );
					
			_t = __t22412;
			_t = _t.getNextSibling();
		}
		catch (RecognitionException ex)
		{
			reportError(ex);
			if (null != _t)
			{
				_t = _t.getNextSibling();
			}
		}
		retTree_ = _t;
	}
	
/** Dump instrumentation for a call statement.
 *  The reference to rule expr prints out the arg
 *  and then at the end of this rule, we close the
 *  generated called to dbg.invoke().
 */
	public void callDumpInstrumentation(AST _t) //throws RecognitionException
{
		
		AST callDumpInstrumentation_AST_in = (AST)_t;
		AST id = null;
		AST e = null;
		
		try {      // for error handling
			AST __t22415 = _t;
			AST tmp31_AST_in = (AST)_t;
			match(_t,CALL);
			_t = _t.getFirstChild();
			id = (AST)_t;
			match(_t,ID);
			_t = _t.getNextSibling();
			Console.Error.Write("{dbg.invoke(\""+id.getText()+"\", \"");
			AST tmp32_AST_in = (AST)_t;
			match(_t,LPAREN);
			_t = _t.getNextSibling();
			{
				if (null == _t)
					_t = ASTNULL;
				switch ( _t.Type )
				{
				case ID:
				case PLUS:
				case STAR:
				case INT:
				{
					e = _t==ASTNULL ? null : (AST)_t;
					expr(_t);
					_t = retTree_;
					break;
				}
				case RPAREN:
				{
					break;
				}
				default:
				{
					throw new NoViableAltException(_t);
				}
				 }
			}
			AST tmp33_AST_in = (AST)_t;
			match(_t,RPAREN);
			_t = _t.getNextSibling();
			AST tmp34_AST_in = (AST)_t;
			match(_t,SEMI);
			_t = _t.getNextSibling();
			Console.Error.Write("\"); ");
			_t = __t22415;
			_t = _t.getNextSibling();
		}
		catch (RecognitionException ex)
		{
			reportError(ex);
			if (null != _t)
			{
				_t = _t.getNextSibling();
			}
		}
		retTree_ = _t;
	}
	
	
	public static readonly string[] tokenNames_ = new string[] {
		@"""<0>""",
		@"""EOF""",
		@"""<2>""",
		@"""NULL_TREE_LOOKAHEAD""",
		@"""CALL""",
		@"""LBRACE""",
		@"""RBRACE""",
		@"""if""",
		@"""then""",
		@"""else""",
		@"""ID""",
		@"""ASSIGN""",
		@"""SEMI""",
		@"""PLUS""",
		@"""STAR""",
		@"""INT""",
		@"""LPAREN""",
		@"""RPAREN""",
		@"""WS""",
		@"""SL_COMMENT""",
		@"""DIGIT"""
	};
	
	private static long[] mk_tokenSet_0_()
	{
		long[] data = { 2224L, 0L};
		return data;
	}
	public static readonly BitSet tokenSet_0_ = new BitSet(mk_tokenSet_0_());
}

